src/mpid/ch4/generic/gpu/mpidig_gpu_utils.h

```diff
@@ -0,0 +1,132 @@
+MPL_STATIC_INLINE_PREFIX void *MPIDIG_gpu_alloc_stage_buf(MPI_Aint count, MPI_Datatype datatype)
+{
+    char *tmp_buf = NULL;
+    size_t size;
+    MPIDI_Datatype_check_size(datatype, count, size);
+    MPL_gpu_malloc_host((void **) &tmp_buf, size);
+
+    return tmp_buf;
+}
+
+/* MPIDIG_gpu_free_stage_buf - free stage buffer */
+MPL_STATIC_INLINE_PREFIX void MPIDIG_gpu_free_stage_buf(void *buf)
+{
+    MPL_gpu_free_host(buf);
+}
+
+/* MPIDIG_gpu_stage_copy_gtoc - moving data from GPU buffer to Stage CPU buffer. */
+MPL_STATIC_INLINE_PREFIX void MPIDIG_gpu_stage_copy_gtoc(const void *sbuf, void *dbuf,
+                                                         MPI_Aint count, MPI_Datatype datatype)
+{
+    int dt_contig;
+    size_t size;
+    MPIR_Datatype *dt_ptr;
+    uintptr_t actual_pack_bytes;
+
+    MPIDI_Datatype_get_size_dt_ptr(count, datatype, size, dt_ptr);
+    dt_contig = (dt_ptr) ? (dt_ptr)->is_contig : 1;
+    if (dt_contig) {
+        /* Stage buffer is contiguous, we can move directly for contiguous datatype. */
+        MPIR_Typerep_pack(sbuf, count, datatype, 0, dbuf, size, &actual_pack_bytes);
+    } else {
+        double density = ((double) (dt_ptr->size)) / ((double) (dt_ptr->max_contig_blocks));
+        if (density >= MPIR_CVAR_CH4_IOV_DENSITY_MIN) {
+            /* No need to pack for high-density datatype */
+            /* We convert it to iov, then copy each iov into stage buffer */
+            int i;
+            MPI_Aint curr_len;
+            void *curr_loc;
+            MPI_Aint num_iov;
+            int n_iov;
+            struct iovec *dt_iov;
+            int actual_iov_len;
+            MPI_Aint actual_iov_bytes;
+            size_t accumulated_count = 0;
+
+            MPIR_Typerep_iov_len(count, datatype, size, &num_iov);
+            n_iov = (int) num_iov;
+            dt_iov = (struct iovec *) MPL_malloc(n_iov * sizeof(struct iovec), MPL_MEM_BUFFER);
+            MPIR_Typerep_to_iov(NULL, count, datatype, 0, dt_iov, n_iov, size, &actual_iov_len,
+                                &actual_iov_bytes);
+            n_iov = actual_iov_len;
+            MPIR_Assert(actual_iov_bytes == (MPI_Aint) size);
+
+            for (i = 0; i < n_iov; i++) {
+                curr_loc = dt_iov[i].iov_base;
+                curr_len = dt_iov[i].iov_len;
+                MPIR_Typerep_pack((char *) sbuf + MPIR_Ptr_to_aint(curr_loc), curr_len, MPI_BYTE, 0,
+                                  (char *) dbuf + accumulated_count, curr_len, &actual_pack_bytes);
+                accumulated_count += curr_len;
+            }
+            MPL_free(dt_iov);
+        } else {
+            /* For low-density datatype, pack into contiguous stage buffer. */
+            MPIR_Typerep_pack(sbuf, count, datatype, 0, dbuf, size, &actual_pack_bytes);
+        }
+    }
+}
+
+/* MPIDIG_gpu_stage_copy_ctog - moving data from Stage CPU buffer to GPU buffer. */
+MPL_STATIC_INLINE_PREFIX void MPIDIG_gpu_stage_copy_ctog(const void *sbuf, void *dbuf,
```

Comment:
Can you use `H2D`, `D2H`, terminology instead of `ctog`?  That'd be consistent with what other GPU APIs use.